pkg/objstore/oss/oss.go

```diff
@@ -0,0 +1,380 @@
+
+	"github.com/go-kit/kit/log"
+	"github.com/pkg/errors"
+	"github.com/thanos-io/thanos/pkg/objstore"
+	"github.com/thanos-io/thanos/pkg/runutil"
+	"gopkg.in/yaml.v2"
+)
+
+type Config struct {
+	Endpoint        string `yaml:"endpoint"`
+	Bucket          string `yaml:"bucket"`
+	AccessKeyID     string `yaml:"access_key_id"`
+	AccessKeySecret string `yaml:"access_key_secret"`
+}
+
+type Bucket struct {
+	name   string
+	logger log.Logger
+	client *alioss.Client
+	config Config
+	bucket *alioss.Bucket
+}
+
+func NewTestBucket(t testing.TB) (objstore.Bucket, func(), error) {
+	c := configFromEnv()
+	err := func(conf Config) error {
+		if conf.Endpoint == "" {
+			return errors.New("no aliyun oss endpoint in config file")
+		}
+		if conf.AccessKeyID == "" {
+			return errors.New("access_key_id is not present in config file")
+		}
+		if conf.AccessKeySecret == "" {
+			return errors.New("access_key_secret is not present in config file")
+		}
+		return nil
+	}(c)
+	if err != nil {
+		return nil, nil, err
+	}
+	if c.Bucket != "" && os.Getenv("THANOS_ALLOW_EXISTING_BUCKET_USE") == "true" {
+		t.Log("ALIYUNOSS_BUCKET is defined. Normally this tests will create temporary bucket " +
+			"and delete it after test. Unset ALIYUNOSS_BUCKET env variable to use default logic. If you really want to run " +
+			"tests against provided (NOT USED!) bucket, set THANOS_ALLOW_EXISTING_BUCKET_USE=true.")
+		return NewTestBucketFromConfig(t, c, true)
+	}
+	return NewTestBucketFromConfig(t, c, false)
+}
+
+func calculateChunks(name string, r io.Reader) (int, int64, error) {
+	switch r.(type) {
+	case *os.File:
+		f, _ := r.(*os.File)
+		if fileInfo, err := f.Stat(); err == nil {
+			s := fileInfo.Size()
+			return int(math.Floor(float64(s) / alioss.MaxPartSize)), s % alioss.MaxPartSize, nil
+		}
+	case *strings.Reader:
+		f, _ := r.(*strings.Reader)
+		return int(math.Floor(float64(f.Size()) / alioss.MaxPartSize)), f.Size() % alioss.MaxPartSize, nil
+	}
+	return -1, 0, errors.New("unsupported implement of io.Reader")
+}
+
+// Upload the contents of the reader as an object into the bucket.
+func (b *Bucket) Upload(ctx context.Context, name string, r io.Reader) error {
+	chunksnum, lastslice, err := calculateChunks(name, r)
+	if err != nil {
+		return err
+	}
+
+	ncloser := ioutil.NopCloser(r)
+	switch chunksnum {
+	case 0:
+		if err := b.bucket.PutObject(name, ncloser); err != nil {
+			return errors.Wrap(err, "failed to upload oss object")
+		}
+	default:
+		{
+			init, err := b.bucket.InitiateMultipartUpload(name)
+			if err != nil {
+				return errors.Wrap(err, "failed to initiate multi-part upload")
+			}
+			chunk := 0
+			uploadEveryPart := func(everypartsize int64, cnk int) (alioss.UploadPart, error) {
+				prt, err := b.bucket.UploadPart(init, ncloser, everypartsize, cnk)
+				if err != nil {
+					if err := b.bucket.AbortMultipartUpload(init); err != nil {
+						return prt, errors.Wrap(err, "failed to upload multi-part chunk")
+					}
+				}
```

Comment:
We are not returning error when UploadPart fails & AbortMultipartUpload succeeds.

The code should be something like:
```
		prt, err := b.bucket.UploadPart(init, ncloser, everypartsize, cnk)
				if errAbort != nil {
					if errAbort := b.bucket.AbortMultipartUpload(init); err != nil {
						return prt, errors.Wrap(err, "failed to upload abort multi-part chunk")
					}
                                       return prt, errors.Wrap(err, "failed to upload multi-part chunk")
				}
```

Reply:
Agree.

Reply:
Good catch!